server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,382 @@
+                                    // no more entries
+                                    assert zipEntry == null;
+                                    assert entryQueue.isEmpty() : entryQueue.size();
+                                    zipOutputStream.finish();
+                                    isPartComplete = true;
+                                    isLastPart = true;
+                                    transferReleasable(releasables);
+                                    assert getNextPart == null;
+                                } else if (zipEntry != null) {
+                                    // new entry, so write the entry header
+                                    zipOutputStream.putNextEntry(zipEntry);
+                                    zipEntry = null;
+                                } else {
+                                    // writing entry body
+                                    if (bodyPart.isPartComplete() == false) {
+                                        try (var innerChunk = bodyPart.encodeChunk(sizeHint, recycler)) {
+                                            final var iterator = innerChunk.iterator();
+                                            BytesRef bytesRef;
+                                            while ((bytesRef = iterator.next()) != null) {
+                                                zipOutputStream.write(bytesRef.bytes, bytesRef.offset, bytesRef.length);
+                                            }
+                                        }
+                                    }
+                                    if (bodyPart.isPartComplete()) {
+                                        if (bodyPart.isLastPart()) {
+                                            zipOutputStream.closeEntry();
+                                            transferReleasable(releasables);
+                                            final var newQueueLength = queueLength.decrementAndGet();
+                                            if (newQueueLength == 0) {
+                                                // next entry isn't available yet, so we stop iterating
+                                                isPartComplete = true;
+                                                assert getNextPart == null;
+                                                getNextPart = continuationListener::addListener;
+                                            } else {
+                                                // next entry is immediately available so start sending its chunks too
+                                                final var nextEntry = entryQueue.poll();
+                                                assert nextEntry != null;
+                                                zipEntry = nextEntry.zipEntry();
+                                                bodyPart = nextEntry.firstBodyPart();
+                                                releasable = nextEntry.releasable();
```

Comment:
I think this should never run concurrently with `enqueueEntry` (while `if` branch can and is OK)? If that is the case, I wonder whether we could assert a single thread access between this code path and `enqueueEntry`?

Reply:
No - I added some commentary about this.

Reply:
> we could assert a single thread access

I'm not sure how, at least not without introducing quite some extra complexity. It's not one actual thread, but there is a happens-before relationship between reads and writes.